libopae/plugins/xfpga/metrics/metrics_utils.c

```diff
@@ -0,0 +1,1231 @@
+
+	}
+	return result;
+}
+
+
+
+fpga_result  enum_bmc_metrics_info(struct _fpga_handle *_handle,
+								fpga_metric_vector *vector,
+								uint64_t *metric_num,
+								enum fpga_hw_type  hw_type)
+{
+	fpga_result result						= FPGA_OK;
+	uint32_t x								= 0;
+	uint32_t num_sensors					= 0;
+	uint32_t num_values						= 0;
+	enum fpga_metric_type metric_type		= FPGA_METRIC_TYPE_POWER;
+	char group_name[SYSFS_PATH_MAX]			= { 0 };
+	char qualifier_name[SYSFS_PATH_MAX]		= { 0 };
+	char units[SYSFS_PATH_MAX]				= { 0 };
+	sdr_details details;
+	bmc_sdr_handle records;
+	bmc_values_handle values;
+
+	if (vector == NULL ||
+		metric_num == NULL) {
+		FPGA_ERR("Invalid input");
+		return result;
+	}
+	result = xfpga_bmcLoadSDRs(_handle, &records, &num_sensors);
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to load BMC SDR.");
+		return result;
+	}
+
+	result = xfpga_bmcReadSensorValues(_handle, records, &values, &num_values);
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to read BMC sensor values.");
+		return result;
+	}
+
+	for (x = 0; x < num_sensors; x++) {
+		result = xfpga_bmcGetSDRDetails(_handle, values, x, &details);
+
+
+		if (details.type == BMC_THERMAL) {
+
+			metric_type = FPGA_METRIC_TYPE_THERMAL;
+			snprintf(group_name, sizeof(group_name), "%s", THERLGMT);
+			snprintf(units, sizeof(units), "%s", TEMP);
+			snprintf_s_ss(qualifier_name, sizeof(qualifier_name), "%s:%s", THERLGMT, details.name);
+
+		} else if (details.type == BMC_POWER) {
+
+			metric_type = FPGA_METRIC_TYPE_POWER;
+			snprintf(group_name, sizeof(group_name), "%s", PWRMGMT);
+			snprintf_s_ss(qualifier_name, sizeof(qualifier_name), "%s:%s", PWRMGMT, details.name);
+			snprintf(units, sizeof(units), "%ls", details.units);
+		} else {
+				continue;
+		}
+
+		result = add_metric_vector(vector, *metric_num, qualifier_name, group_name, "", details.name, "", units, FPGA_METRIC_DATATYPE_DOUBLE, metric_type, hw_type, 0);
+		if (result != FPGA_OK) {
+			FPGA_MSG("Failed to add metrics");
+			return result;
+		}
+
+		*metric_num = *metric_num + 1;
+	}
+
+	result = xfpga_bmcDestroySensorValues(_handle, &values);
+	if (result != FPGA_OK) {
+		FPGA_MSG("Failed to Destroy Sensor value.");
+	}
+
+	result = xfpga_bmcDestroySDRs(_handle, &records);
+	if (result != FPGA_OK) {
+		FPGA_ERR("Failed to Destroy SDR.");
+		return result;
+	}
+
+	return result;
+}
+
+fpga_result free_fpga_enum_metrics_vector(struct _fpga_handle *_handle)
+{
+	fpga_result result = FPGA_OK;
+	uint64_t i = 0;
+
+	if (_handle == NULL) {
+		FPGA_ERR("Invalid handle ");
+		return FPGA_INVALID_PARAM;
+	}
+
+	if (_handle->magic != FPGA_HANDLE_MAGIC) {
+		FPGA_MSG("Invalid handle");
+		return FPGA_INVALID_PARAM;
+	}
+
+
+	for (i = 0; i < fpga_vector_total(&(_handle->fpga_enum_metric_vector)); i++) {
+		fpga_vector_delete(&(_handle->fpga_enum_metric_vector), i);
+	}
+
+	fpga_vector_free(&(_handle->fpga_enum_metric_vector));
+
+	if (_handle->dl_handle) {
```

Comment:
Need to change the name of dl_handle to be something more-specific to metrics.  This name is too generic.